Refactored Step Builder Messages

bradenmacdonald opened the pull request:
This refactors step builder to:
* Make the Review step code simpler
* Split the parts of the review screen up into XBlocks that are more flexible and re-orderable.
* Make the Step Builder block preview/work in Studio as students would see it in the LMS

**Example: Editing a review step**:
![screen shot 2015-11-01 at 8 36 53 pm](https://cloud.githubusercontent.com/assets/945577/10874391/58fc6730-80d8-11e5-869c-445d7bfd8739.png)


**Example: A Conditional Message**:
![screen shot 2015-11-01 at 8 06 24 pm](https://cloud.githubusercontent.com/assets/945577/10874364/cee4e34c-80d7-11e5-82ea-57f7c08b2130.png)

smarnach commented on problem_builder/mentoring.py:
```diff
@@ -1101,9 +1115,7 @@ def try_again(self, data, suffix=''):
-        context = context.copy() if context else {}
-        context['author_preview_view'] = True
-        return super(MentoringWithExplicitStepsBlock, self).author_preview_view(context)
+        return self.student_view(context)
```
Isn't it the implicit behaviour to fall back to the student view?  (No need to change anything, it's fine to keep this explicit in any case.  I'm just curious.)

bradenmacdonald replied:
Yes, but if it renders `student_view` as a fallback, Studio will wrap the XBlock in a bunch of HTML (header, edit controls) that we don't want. Studio looks at the name of the view that's being rendered when deciding to wrap the element or not.

smarnach commented on problem_builder/public/js/mentoring_with_steps.js:
```diff
@@ -8,23 +8,22 @@ function MentoringWithStepsBlock(runtime, element) {
-        gradeDOM, attemptsDOM, reviewTipsDOM, reviewLinkDOM, submitXHR;
+    var checkmark, submitDOM, nextDOM, reviewButtonDOM, tryAgainDOM,
+        gradeDOM, attemptsDOM, reviewLinkDOM, submitXHR;
+    var reviewStepDOM = $("[data-block-type=sb-review-step]", element);
```
It's probably more efficient to restrict this a bit further, like `"div.xblock[data-block-type=sb-review-step]"`.  Otherwise jQuery will have to iterate through all DOM objects and retrieve the `data-block-type` attribute.  (Not sure whether what I suggested would be correct in all cases.  Also not sure whether performance matters here.)

bradenmacdonald replied:
I agree about the "DOM" name in the variables, but that's an existing convention in this code.

Also yeah, the problem with the selector is that in Workbench the XBlock DIVs have the class `xblock-v1` but in Studio they have the class `xblock` :-/ I can just double the selector though.

smarnach commented on problem_builder/public/js/mentoring_with_steps.js:
```diff
@@ -111,7 +93,8 @@ function MentoringWithStepsBlock(runtime, element) {
                 // Update the number of attempts.
                 attemptsDOM.data('num_attempts', response.num_attempts);
-                updateGrade(response.grade_data);
+                reviewStepDOM.html($(response.review_html).html());
```
Again a question for my personal education: How is this different from `reviewStepDOM.html(response.review_html)`?  Does it just remove the outermost HTML tag?  If so, why is it included in `review_html` in the first place?

bradenmacdonald replied:
@smarnach Yes, it removes the outermost HTML tag. It's included because the XBlock runtime puts it there and we can't control that. (I could bypass the normal block rendering pipeline which wraps it in that DIV but then we would miss other important steps in the render pipeline, such as link rewriting).

This could in theory be simplified to `reviewStepDOM.replaceWith($(response.review_html))` but that would invalidate the jQuery object / DOM reference `reviewStepDOM`, so I found this to be cleaner.

smarnach commented on problem_builder/step_review.py:
```diff
@@ -0,0 +1,291 @@
+    def student_view(self, context=None):
+        """ Render this message. """
+        html = u'<div class="review-conditional-message">{content}</div>'.format(
+            content=self.content
```
Do we need HTML escaping here?

bradenmacdonald replied:
No, problem builder blocks generally allow the authors to enter custom HTML.

smarnach commented on problem_builder/step_review.py:
```diff
@@ -0,0 +1,291 @@
+
+    def student_view(self, context=None):
+        """ Render the score summary message. """
+        html = loader.render_template("templates/html/sb-review-score.html", context.get("score_summary", {}))
```
This won't work when `context` has its default value `None`.

smarnach commented on problem_builder/step_review.py:
```diff
@@ -0,0 +1,291 @@
+    mentoring_view = student_view  # Same as student_view but Studio won't wrap it with the editing header/buttons
+
+    def author_view(self, context=None):
+        if not context.get("score_summary"):
```
Same here.  (Not sure whether that ever happens, but then I wonder why we have a default value for `context` anyway.)

smarnach commented on problem_builder/step_review.py:
```diff
@@ -0,0 +1,291 @@
+@XBlock.needs("i18n")
+class PerQuestionFeedbackBlock(XBlockWithTranslationServiceMixin, XBlockWithPreviewMixin, XBlock):
+    """
+    Summaryize the score that the student earned.
```
Spelling: "Summaryize"

smarnach commented on problem_builder/step_review.py:
```diff
@@ -0,0 +1,291 @@
+
+    def author_view(self, context=None):
+        """ Show example content in Studio """
+        if not context.get("per_question_review_tips"):
```
Check for `context is None` missing again.

smarnach commented on problem_builder/templates/html/sb-review-per-question-feedback.html:
```diff
@@ -0,0 +1,9 @@
+{% load i18n %}
+
+<!-- Tips about specific questions the student got wrong. From pb-message[type=on-assessment-review-question] blocks -->
```
This comment will be included in the resulting HTML.  You can use

    {# Your comment here. #}

in Django templates to avoid that.

smarnach commented:
@bradenmacdonald Nice, this makes the code indeed more readable!  I've done a first past through the code, but haven't done any testing yet.  To facilitate getting this merged until tomorrow morning, I'll do the  testing after the meeting, so if anything comes up you will still have a bit of time to address any issues.

How much time will you need for the tests?  If you can finish them soon, I might be able to give a thumbs up today.  (I'll be available until about 22:00 UTC.)

bradenmacdonald commented:
@smarnach Thanks for the initial review! Yes I'm happy with how this has improved the code :-)

I'll work on addressing your comments and updating the tests right after the meeting, but I'm not sure how long it will take. A couple hours perhaps?

smarnach commented:
@bradenmacdonald OK, we'll see then.  Either I'm still available to take a final look, or it will have to wait until tomorrow.

bradenmacdonald commented:
@smarnach  I've addressed your initial review comments and fixed a small bug. Now I'm just continuing to get the tests passing again.

smarnach commented:
When adding a new Per-Question Feedback block and cliking "Edit" on it, I get this error:

![screenshot_2015-11-02_22-07-52](https://cloud.githubusercontent.com/assets/249196/10894312/33fd3ce4-81ae-11e5-9131-e7b5e016066d.png)

(This is on a relatively current version of the solutions devstack.)

bradenmacdonald commented:
@smarnach Good catch. Fix just pushed: ee49a27365576954b27adf71c68935b084595733

(I also just pushed two other fixes and will "soon" push a commit that makes the tests pass, hopefully)

smarnach commented:
Does the new version have to be backwards compatible to existing course data?  When loading my pre-existing Step Builder example in Studio, one of the messages was displayed with the heading "INVALID MESSAGE".  I also still had the old "Message shown when complete/incomplete" blocks, and things didn't properly work when tested in the LMS.  I'll try again with a newly created version.

bradenmacdonald commented:
@smarnach I'm aware of that, but I think it's fine. I don't believe this is yet in use in any courses. The important thing is that the authors are able to do the conversion.

smarnach commented:
If you add multilple conditional messages and view them from the Step Builder top level in studio, it's hard to see which condition belongs to which message, since the condition is separate from the message by a horizontal rule, while the messages aren't separated from each other at all:

![screenshot_2015-11-02_22-26-45](https://cloud.githubusercontent.com/assets/249196/10894878/4348d8d6-81b1-11e5-901d-70cc98be2f6e.png)

smarnach commented:
I reviewed everything that's there up to now.  The code is looking good, and everything is working fine with the latest changes as well.  So the only thing left is to make the tests pass. :)

I'll take a final look tomorrow morning, and will merge and do the hash update on the Solutions fork of edx-platform if everything looks good.

bradenmacdonald commented:
Thanks a lot, @smarnach :)

smarnach commented:
All looking good, :+1:.
